### PR TITLE
Disable format parsing and track parsed fields

### LINKED ISSUE #3282: Parse is added after transform
The parsing should be done before the aggregate transform. Note that this is not true in general for all parses. The parse of `max_price` should happen after the aggregation. 

```json
{
  "$schema": "https://vega.github.io/schema/vega-lite/v2.json",
  "width": 600,
  "height": 300,
  "data": {
    "url": "data/stocks.csv",
    "format": {"parse": {"price": "number"}}
  },
  "transform": [{
    "aggregate": [{"op": "max", "field": "price", "as": "max_price"}],
    "groupby": ["date"]
  }],
  "mark": "point",
  "encoding": {
    "x": {"field": "date", "type": "temporal", "axis": {"format": "%Y"}},
    "y": {"field": "max_price", "type": "quantitative"}
  }
}
```

### LINKED ISSUE #3665: Disable format parse
Please:
- [x] Check for duplicate issues. Please file separate requests as separate issues on GitHub.
- [x] Describe how to reproduce the bug / the goal of the feature request.
- [x] Use the latest versions of Vega and Vega-Lite.
- [x] Provide an example spec in JSON and, if applicable, screenshots, GIF videos (e.g. using [LICEcap](https://www.cockos.com/licecap/)), or working example (e.g. based on [this Block](https://bl.ocks.org/domoritz/455e1c7872c4b38a58b90df0c3d7b1b9)).
- [x] Please ensure all code snippets and error messages are formatted in appropriate code blocks. See [Creating and highlighting code blocks](https://help.github.com/articles/creating-and-highlighting-code-blocks/).

Is it possible to get vega lite to compile vega that doesn't try to transform the data to the right type? For example, in [the bar example](https://vega.github.io/editor/#/examples/vega-lite/bar), it has this transform:

```json
    {
      "name": "data_0",
      "source": "source_0",
      "transform": [
        {"type": "formula", "expr": "toNumber(datum[\"b\"])", "as": "b"},
        {
          "type": "filter",
          "expr": "datum[\"b\"] !== null && !isNaN(datum[\"b\"])"
        }
      ]
    }
```

I know I can remove the `null` checks by setting `invalidValues` to `null` in the config, but I am still left with the `toNumber` conversion. What if I already know it will be a number, isn't this conversion superfluous? Is there a way to disable it and have the vega that is produced only create one data source?

I am asking because [the mapd implementation of transform](https://www.mapd.com/docs/latest/mapd-core-guide/vegaReference/#vegarefdatatrans) only supports `formula` transforms against previously aggregated columns and [I am trying to render vega lite with mapd](https://github.com/Quansight/jupyter-mapd-renderer/pull/7).

### PR DESCRIPTION
**Pull request by domoritz**
Fixes #3665 
Fixes #3282

CC @saulshanabrook

### CODE REVIEW

File: src/data.ts [hunk 1]
```diff
@@ -3,16 +3,21 @@
...
+   * If set to `"auto"` (the default), perform automatic type inference to determine the desired data types.
+   * If set to `null`, disable type inference based on the spec and only use type inference based on the data.
+   * Alternatively, a parsing directive object can be provided for explicit data types. Each property of the object corresponds to a field name, and the value to the desired data type (one of `"number"`, `"boolean"`, `"date"`, or null (do not parse the field)).
    * For example, `"parse": {"modified_on": "date"}` parses the `modified_on` field in each input record a Date value.
    *
    * For `"date"`, we parse data based using Javascript's [`Date.parse()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/parse).
    * For Specific date formats can be provided (e.g., `{foo: 'date:"%m%d%Y"'}`), using the [d3-time-format syntax](https://github.com/d3/d3-time-format#locale_format). UTC date format parsing is supported similarly (e.g., `{foo: 'utc:"%m%d%Y"'}`). See more about [UTC time](timeunit.html#utc)
    */
-  parse?: 'auto' | object;
+  parse?: 'auto' | Parse | null;
```
Comment (kanitw) [inline src/data.ts]: What's the behavior for  providing an object for only a subset of fields?

(Do we do "auto" for all other fields unmentioned?)

**Comment by domoritz** [comment]
I updated the PR to properly track parsed fields. This is something we have not done correctly until now.

File: src/compile/data/formatparse.ts [hunk 2]
```diff
@@ -52,77 +53,85 @@ export class ParseNode extends DataFlowNode {
...
+          log.warn(log.message.differentParse(field, explicit[field], parsedAs));
+        }
       }
     }
 
-    if (keys(parse).length === 0) {
+    const parse = new Split(explicit, implicit);
+
+    // add the format parse from this model so that children don't parse the same field again
+    ancestorParse.copyAll(parse);
```
Comment (kanitw) [inline src/compile/data/formatparse.ts]: So you merge this into parent, but still create a local parse node? 

I'm pretty confused.

File: src/compile/data/index.ts [hunk 3]
```diff
@@ -43,7 +44,23 @@ export interface DataComponent {
...
 
   /**
-   * Parse properties passed down from ancestors.
+   * Parse properties passed down from ancestors. Helps us to keep track of what has been parsed or is derived.
    */
-  ancestorParse: Dict<string>;
+  ancestorParse?: AncestorParse;
+}
+
+export class AncestorParse extends Split<Dict<string>> {
```
Comment (kanitw) [inline src/compile/data/index.ts]: Why call it AncestorParse?  I don't see anything that is specific about Ancestor. 

ParseComponent is probably more appropriate and consistent with the rest of the code base.

File: src/compile/data/parse.ts [hunk 4]
```diff
@@ -160,8 +191,15 @@ Formula From Sort Array
 export function parseData(model: Model): DataComponent {
   let head = parseRoot(model, model.component.data.sources);
 
-  const outputNodes = model.component.data.outputNodes;
-  const outputNodeRefCounts = model.component.data.outputNodeRefCounts;
+  const {outputNodes, outputNodeRefCounts} = model.component.data;
+  const ancestorParse = model.parent ? model.parent.component.data.ancestorParse.clone() : new AncestorParse();
```
Comment (kanitw) [inline src/compile/data/parse.ts]: I don't understand why do you have to clone here?

File: src/predicate.ts [hunk 5]
```diff
@@ -32,9 +32,7 @@ export function isSelectionPredicate(predicate: LogicalOperand<Predicate>): pred
   return predicate && predicate['selection'];
 }
 
-export interface FieldEqualPredicate {
-  // TODO: support aggregate
-
+export interface FieldPredicateBase {
```
Comment (kanitw) [inline src/predicate.ts]: This is a nice refactor, but seems irrelvant to this PR.

File: src/compile/data/formatparse.ts [hunk 6]
```diff
@@ -52,77 +53,85 @@ export class ParseNode extends DataFlowNode {
...
-        keys(p).forEach(field => {
-          parse[field] = p[field];
-        });
+    return this.makeWithAncestors(parent, {}, implicit, ancestorParse);
+  }
+
+  /**
+   * Creates a parse node from two sets of explicit and implicit parse and updates ancestorParse.
+   */
+  public static makeWithAncestors(parent: DataFlowNode, explicit: Dict<string>, implicit: Dict<string>, ancestorParse: AncestorParse) {
```
Comment (kanitw) [inline src/compile/data/formatparse.ts]: Why do you need to pass `ancestorParse` here?   Shouldn't it be a part of `parent`?  

(Sorry this is very complicated so it's hard to get why.)

File: src/compile/data/formatparse.ts (hunk 2 above)
Comment (domoritz) [reply src/compile/data/formatparse.ts]: ancestorParse is not a parse node, it's an object to track what has been parsed already

File: src/compile/data/index.ts (hunk 3 above)
Comment (domoritz) [reply src/compile/data/index.ts]: This already existed: https://github.com/vega/vega-lite/blob/master/src/compile/data/index.ts#L48. Let's rename things in a new PR.

File: src/predicate.ts (hunk 5 above)
Comment (domoritz) [reply src/predicate.ts]: I had to use it at some point but then it ended up being automatically inferred by Typescript.

File: src/compile/data/formatparse.ts [hunk 7]
```diff
@@ -52,77 +53,85 @@ export class ParseNode extends DataFlowNode {
...
-        return null;  // disable parsing completely
-      } else if (p) {
-        keys(p).forEach(field => {
-          parse[field] = p[field];
-        });
+    return this.makeWithAncestors(parent, {}, implicit, ancestorParse);
+  }
+
+  /**
+   * Creates a parse node from two sets of explicit and implicit parse and updates ancestorParse.
```
[the remark this answers is not in the document]
Comment (domoritz) [reply src/compile/data/formatparse.ts]: explicit = data.format.parse
implicit = inferred from transforms and field types

File: src/compile/data/formatparse.ts (hunk 6 above)
Comment (domoritz) [reply src/compile/data/formatparse.ts]: No, the ancestor parse applies to the current model as well. I could simplify this and not make a copy, I guess.

File: src/compile/data/formatparse.ts [hunk 8]
```diff
@@ -52,77 +53,85 @@ export class ParseNode extends DataFlowNode {
...
 
-    // We should not parse what has already been parsed in a parent
-    const modelParse = model.component.data.ancestorParse;
-    keys(modelParse).forEach(field => {
-      if (parse[field] !== modelParse[field]) {
-        log.warn(log.message.differentParse(field, parse[field], modelParse[field]));
+    for (const field of keys(explicit)) {
+      const parsedAs = ancestorParse.get(field);
+      if (parsedAs !== undefined) {
+        if (parsedAs === explicit[field]) {
```
[the remark this answers is not in the document]
Comment (domoritz) [reply src/compile/data/formatparse.ts]: Done

File: src/compile/data/parse.ts (hunk 4 above)
Comment (domoritz) [reply src/compile/data/parse.ts]: I don't have to. I guess we could just reuse the same object and modify it as we go down the tree.

File: src/compile/split.ts [hunk 9]
```diff
@@ -68,6 +68,17 @@ export class Split<T extends object> {
       this.set(key, s[key], true);
     }
   }
+
+  /**
+   * Merge split object this split object. Properties from the other split
```
[the remark this answers is not in the document]
Comment (domoritz) [reply src/compile/split.ts]: Done

File: src/compile/split.ts [hunk 10]
```diff
@@ -68,6 +68,17 @@ export class Split<T extends object> {
...
+   * overwrite properties from this split.
+   */
+  public copyAll(other: Split<T>) {
```
[the remark this answers is not in the document]
Comment (domoritz) [reply src/compile/split.ts]: I looked at `copyKeyFromObject` and `copyKeyFromObject` in the same class. What do you suggest? 

Yes, I want side effects because I don't want to have to return a copy everywhere.

**Comment by domoritz** [comment]
Hmm, the last commit broke some things. Looking into it.

File: src/compile/data/index.ts (hunk 3 above)
Comment (kanitw) [reply src/compile/data/index.ts]: Your link only shows an old `Dict<string>`, but this class is introduced in this PR so naming it property should be in the scope of this PR.

File: src/compile/data/formatparse.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (kanitw) [reply src/compile/data/formatparse.ts]: The above part has

> // We always ignore derived fields even if they are implicitly defined because we expects users to create the right types.

The logic for this part is different, but doesn't have any comment

File: src/compile/data/formatparse.ts (hunk 6 above)
Comment (kanitw) [reply src/compile/data/formatparse.ts]: If you can, better do it now.

File: src/compile/data/formatparse.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (kanitw) [reply src/compile/data/formatparse.ts]: Better remove "two sets of". 

"two sets of explicit and implicit parse" sounds like you have 2 x explicit parse and 2 x implicit parse

File: src/compile/data/formatparse.ts (hunk 6 above)
Comment (domoritz) [reply src/compile/data/formatparse.ts]: Actually, that's incorrect when we have forks because every subtree needs to track parsed fields separately.

File: src/compile/data/formatparse.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (domoritz) [reply src/compile/data/formatparse.ts]: Because the logic isn't special here. If a field was parsed explicitly, we don't ignore what it has been parsed as.

File: src/compile/data/index.ts (hunk 3 above)
Comment (domoritz) [reply src/compile/data/index.ts]: ParseComponent is not more appropriate, though because we never generate a parse object from this. I added a comment to make this clear.

File: src/compile/data/parse.ts (hunk 4 above)
Comment (domoritz) [reply src/compile/data/parse.ts]: I was wrong. We have to as explained earlier.

File: src/compile/data/formatparse.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (kanitw) [reply src/compile/data/formatparse.ts]: There are two cases -- each of them is different.  So better comment so it's easier for reader.

File: src/compile/data/parse.ts (hunk 4 above)
Comment (kanitw) [reply src/compile/data/parse.ts]: > We have to as explained earlier.

Can you explain again?  I don't know where you explain it as this PR is pretty complicated.

The rest of the "parse" codebase produce something as as part of component. 

Isn't it weird that this `ancestorParse` seems to be just floating within the parse codebase?

Comment (kanitw) [reply src/compile/data/parse.ts]: At very least, this line deserves a better comment why this is being cloned.

File: src/compile/data/index.ts [hunk 11]
```diff
@@ -43,7 +44,28 @@ export interface DataComponent {
...
   /**
-   * Parse properties passed down from ancestors.
+   * Parse properties passed down from ancestors. Helps us to keep track of what has been parsed or is derived.
    */
-  ancestorParse: Dict<string>;
+  ancestorParse?: AncestorParse;
+}
+
+/**
+ * Class to track interesting properties (see https://15721.courses.cs.cmu.edu/spring2016/papers/graefe-ieee1995.pdf)
```
Comment (kanitw) [inline src/compile/data/index.ts]: Is this paper really needed?  It's another 8 page people have to read. 

Better provide a TLDR explanation.

File: src/compile/data/index.ts [hunk 12]
```diff
@@ -43,7 +44,28 @@ export interface DataComponent {
...
+ * about how fields have been parsed or whether they have been derived in a transforms. We use this to not parse the
+ * same field again (or differently).
+ */
+export class AncestorParse extends Split<Dict<string>> {
```
Comment (kanitw) [inline src/compile/data/index.ts]: If you want to rename it later, better rename it in this PR as this class is introduced in the PR.

File: src/compile/data/formatparse.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (domoritz) [reply src/compile/data/formatparse.ts]: ok

File: src/compile/data/formatparse.ts (hunk 6 above)
Comment (kanitw) [reply src/compile/data/formatparse.ts]: I don't understand your explanation here yet... 

Please explain "Why do you need to pass ancestorParse here?"

File: src/compile/data/index.ts (hunk 12 above)
Comment (domoritz) [reply src/compile/data/index.ts]: I like the name.

File: src/compile/data/index.ts (hunk 11 above)
Comment (domoritz) [reply src/compile/data/index.ts]: This goes into the depths of query optimization.

File: src/compile/data/formatparse.ts (hunk 6 above)
Comment (domoritz) [reply src/compile/data/formatparse.ts]: Oh, my response was to making a copy.

`ancestorParse` is currently not in nodes but passed as a separate object. I was thinking about adding this information to the nodes themselves but that makes a lot of the merging logic more complicated. I created an issue for this in https://github.com/vega/vega-lite/issues/3715

File: src/compile/data/parse.ts [hunk 13]
```diff
@@ -48,56 +49,81 @@ function parseRoot(model: Model, sources: Dict<SourceNode>): DataFlowNode {
...
+            parse[filter.field] = 'date';
+          }
         }
-      }
+      });
 
-      if (keys(parse).length > 0) {
-        parent = new ParseNode(parent, parse);
+      if (keys(parse).length) {
+        parent = ParseNode.makeWithAncestors(parent, {}, parse, ancestorParse) || parent;
```
Comment (kanitw) [inline src/compile/data/parse.ts]: I find the name `makeImplicit` a bit misleading since this parse from transform is "implicit" too, but doesn't call that `makeImplicit` method.  

May be rename `makeImplicit` to be`makeImplicitFromEncoding` and extract the logic above here to be `makeImplicitFromFilterTransform` (or just FromTransform).   This way you can make `makeWithAncestor` a `private` method too. 


One question that I have is don't we have to do the same thing with TimeUnit and Bin, in addition to Filter?
